File: wrappers/fsharp/Encoding.fs
module TurboToken.Encoding

open System
open System.Runtime.InteropServices
open System.Text
open TurboToken.Native
open TurboToken.Chat
open TurboToken.Registry

type Encoding =
    { Name: string
      Spec: EncodingSpec
      RankPayload: byte[] }

let private utf8 = Encoding.UTF8

let private twoPassUInt32
    (rankPayload: byte[])
    (input: byte[])
    (fn: nativeint -> unativeint -> nativeint -> unativeint -> nativeint -> unativeint -> nativeint)
    : Result<uint32[], string> =
    let rankHandle = GCHandle.Alloc(rankPayload, GCHandleType.Pinned)
    let inputHandle = GCHandle.Alloc(input, GCHandleType.Pinned)
    try
        let rankPtr = rankHandle.AddrOfPinnedObject()
        let rankLen = unativeint rankPayload.Length
        let inputPtr = inputHandle.AddrOfPinnedObject()
        let inputLen = unativeint input.Length

        let sizeResult = fn rankPtr rankLen inputPtr inputLen IntPtr.Zero (unativeint 0)
        let count = int64 sizeResult
        if count < 0L then
            Error (sprintf "FFI size query returned error code %d" count)
        elif count = 0L then
            Ok Array.empty
        else
            let buffer = Array.zeroCreate<uint32> (int count)
            let outHandle = GCHandle.Alloc(buffer, GCHandleType.Pinned)
            try
                let outPtr = outHandle.AddrOfPinnedObject()
                let written = fn rankPtr rankLen inputPtr inputLen outPtr (unativeint count)
                let writtenCount = int64 written
                if writtenCount < 0L then
                    Error (sprintf "FFI fill returned error code %d" writtenCount)
                elif writtenCount < count then
                    Ok (Array.sub buffer 0 (int writtenCount))
                else
                    Ok buffer
            finally
                outHandle.Free()
    finally
        rankHandle.Free()
        inputHandle.Free()

let private twoPassUInt8
    (rankPayload: byte[])
    (input: uint32[])
    (fn: nativeint -> unativeint -> nativeint -> unativeint -> nativeint -> unativeint -> nativeint)
    : Result<byte[], string> =
    let rankHandle = GCHandle.Alloc(rankPayload, GCHandleType.Pinned)
    let inputHandle = GCHandle.Alloc(input, GCHandleType.Pinned)
    try
        let rankPtr = rankHandle.AddrOfPinnedObject()
        let rankLen = unativeint rankPayload.Length
        let inputPtr = inputHandle.AddrOfPinnedObject()
        let inputLen = unativeint input.Length

        let sizeResult = fn rankPtr rankLen inputPtr inputLen IntPtr.Zero (unativeint 0)
        let count = int64 sizeResult
        if count < 0L then
            Error (sprintf "FFI size query returned error code %d" count)
        elif count = 0L then
            Ok Array.empty
        else
            let buffer = Array.zeroCreate<byte> (int count)
            let outHandle = GCHandle.Alloc(buffer, GCHandleType.Pinned)
            try
                let outPtr = outHandle.AddrOfPinnedObject()
                let written = fn rankPtr rankLen inputPtr inputLen outPtr (unativeint count)
                let writtenCount = int64 written
                if writtenCount < 0L then
                    Error (sprintf "FFI fill returned error code %d" writtenCount)
                elif writtenCount < count then
                    Ok (Array.sub buffer 0 (int writtenCount))
                else
                    Ok buffer
            finally
                outHandle.Free()
    finally
        rankHandle.Free()
        inputHandle.Free()

let encode (enc: Encoding) (text: string) : Result<uint32[], string> =
    let textBytes = utf8.GetBytes(text)
    twoPassUInt32 enc.RankPayload textBytes (fun rp rl ip il op oc ->
        turbotoken_encode_bpe_from_ranks(rp, rl, ip, il, op, oc))

let decode (enc: Encoding) (tokens: uint32[]) : Result<string, string> =
    twoPassUInt8 enc.RankPayload tokens (fun rp rl ip il op oc ->
        turbotoken_decode_bpe_from_ranks(rp, rl, ip, il, op, oc))
    |> Result.map utf8.GetString

let count (enc: Encoding) (text: string) : Result<int, string> =
    let textBytes = utf8.GetBytes(text)
    let rankHandle = GCHandle.Alloc(enc.RankPayload, GCHandleType.Pinned)
    let textHandle = GCHandle.Alloc(textBytes, GCHandleType.Pinned)
    try
        let result =
            turbotoken_count_bpe_from_ranks(
                rankHandle.AddrOfPinnedObject(), unativeint enc.RankPayload.Length,
                textHandle.AddrOfPinnedObject(), unativeint textBytes.Length)
        let code = int64 result
        if code < 0L then Error (sprintf "count returned error code %d" code)
        else Ok (int code)
    finally
        rankHandle.Free()
        textHandle.Free()

let countTokens (enc: Encoding) (text: string) : Result<int, string> =
    count enc text

let isWithinTokenLimit (enc: Encoding) (text: string) (limit: int) : Result<int option, string> =
    let textBytes = utf8.GetBytes(text)
    let rankHandle = GCHandle.Alloc(enc.RankPayload, GCHandleType.Pinned)
    let textHandle = GCHandle.Alloc(textBytes, GCHandleType.Pinned)
    try
        let result =
            turbotoken_is_within_token_limit_bpe_from_ranks(
                rankHandle.AddrOfPinnedObject(), unativeint enc.RankPayload.Length,
                textHandle.AddrOfPinnedObject(), unativeint textBytes.Length,
                unativeint limit)
        let code = int64 result
        if code = -2L then Ok None
        elif code < 0L then Error (sprintf "isWithinTokenLimit returned error code %d" code)
        else Ok (Some (int code))
    finally
        rankHandle.Free()
        textHandle.Free()

let encodeChat (enc: Encoding) (messages: ChatMessage list) (options: ChatOptions) : Result<uint32[], string> =
    let text = formatChat messages options
    encode enc text

let countChat (enc: Encoding) (messages: ChatMessage list) (options: ChatOptions) : Result<int, string> =
    let text = formatChat messages options
    count enc text

let isChatWithinTokenLimit (enc: Encoding) (messages: ChatMessage list) (limit: int) (options: ChatOptions) : Result<int option, string> =
    let text = formatChat messages options
    isWithinTokenLimit enc text limit

let encodeFilePath (enc: Encoding) (path: string) : Result<uint32[], string> =
    let pathBytes = utf8.GetBytes(path)
    twoPassUInt32 enc.RankPayload pathBytes (fun rp rl ip il op oc ->
        turbotoken_encode_bpe_file_from_ranks(rp, rl, ip, il, op, oc))

let countFilePath (enc: Encoding) (path: string) : Result<int, string> =
    let pathBytes = utf8.GetBytes(path)
    let rankHandle = GCHandle.Alloc(enc.RankPayload, GCHandleType.Pinned)
    let pathHandle = GCHandle.Alloc(pathBytes, GCHandleType.Pinned)
    try
        let result =
            turbotoken_count_bpe_file_from_ranks(
                rankHandle.AddrOfPinnedObject(), unativeint enc.RankPayload.Length,
                pathHandle.AddrOfPinnedObject(), unativeint pathBytes.Length)
        let code = int64 result
        if code < 0L then Error (sprintf "countFilePath returned error code %d" code)
        else Ok (int code)
    finally
        rankHandle.Free()
        pathHandle.Free()

let isFilePathWithinTokenLimit (enc: Encoding) (path: string) (limit: int) : Result<int option, string> =
    let pathBytes = utf8.GetBytes(path)
    let rankHandle = GCHandle.Alloc(enc.RankPayload, GCHandleType.Pinned)
    let pathHandle = GCHandle.Alloc(pathBytes, GCHandleType.Pinned)
    try
        let result =
            turbotoken_is_within_token_limit_bpe_file_from_ranks(
                rankHandle.AddrOfPinnedObject(), unativeint enc.RankPayload.Length,
                pathHandle.AddrOfPinnedObject(), unativeint pathBytes.Length,
                unativeint limit)
        let code = int64 result
        if code = -2L then Ok None
        elif code < 0L then Error (sprintf "isFilePathWithinTokenLimit returned error code %d" code)
        else Ok (Some (int code))
    finally
        rankHandle.Free()
        pathHandle.Free()

File: wrappers/cpp/src/encoding.cpp
#include "turbotoken/encoding.hpp"
#include "turbotoken/error.hpp"
#include "turbotoken/rank_cache.hpp"

#include <turbotoken.h>

#include <fstream>
#include <sstream>

namespace turbotoken {

Encoding::Encoding(std::vector<uint8_t> rank_payload, EncodingSpec spec)
    : rank_payload_(std::move(rank_payload)), spec_(std::move(spec)) {}

Encoding Encoding::get(const std::string& name) {
    auto spec = get_encoding_spec(name);
    auto payload = read_rank_file(name);
    return Encoding(std::move(payload), std::move(spec));
}

Encoding Encoding::for_model(const std::string& model) {
    auto enc_name = model_to_encoding(model);
    return get(enc_name);
}

std::vector<uint32_t> Encoding::encode(std::string_view text) const {
    ptrdiff_t n = turbotoken_encode_bpe_from_ranks(
        rank_payload_.data(), rank_payload_.size(),
        reinterpret_cast<const uint8_t*>(text.data()), text.size(),
        nullptr, 0);
    if (n < 0) {
        throw EncodingError("turbotoken_encode_bpe_from_ranks failed");
    }

    std::vector<uint32_t> tokens(static_cast<size_t>(n));
    ptrdiff_t written = turbotoken_encode_bpe_from_ranks(
        rank_payload_.data(), rank_payload_.size(),
        reinterpret_cast<const uint8_t*>(text.data()), text.size(),
        tokens.data(), tokens.size());
    if (written < 0) {
        throw EncodingError("turbotoken_encode_bpe_from_ranks failed");
    }
    tokens.resize(static_cast<size_t>(written));
    return tokens;
}

std::string Encoding::decode(const std::vector<uint32_t>& tokens) const {
    ptrdiff_t n = turbotoken_decode_bpe_from_ranks(
        rank_payload_.data(), rank_payload_.size(),
        tokens.data(), tokens.size(),
        nullptr, 0);
    if (n < 0) {
        throw DecodingError("turbotoken_decode_bpe_from_ranks failed");
    }

    std::string result(static_cast<size_t>(n), '\0');
    ptrdiff_t written = turbotoken_decode_bpe_from_ranks(
        rank_payload_.data(), rank_payload_.size(),
        tokens.data(), tokens.size(),
        reinterpret_cast<uint8_t*>(result.data()), result.size());
    if (written < 0) {
        throw DecodingError("turbotoken_decode_bpe_from_ranks failed");
    }
    result.resize(static_cast<size_t>(written));
    return result;
}

size_t Encoding::count(std::string_view text) const {
    ptrdiff_t n = turbotoken_count_bpe_from_ranks(
        rank_payload_.data(), rank_payload_.size(),
        reinterpret_cast<const uint8_t*>(text.data()), text.size());
    if (n < 0) {
        throw EncodingError("turbotoken_count_bpe_from_ranks failed");
    }
    return static_cast<size_t>(n);
}

size_t Encoding::count_tokens(std::string_view text) const {
    return count(text);
}

std::optional<size_t> Encoding::is_within_token_limit(
    std::string_view text, size_t limit) const {
    ptrdiff_t r = turbotoken_is_within_token_limit_bpe_from_ranks(
        rank_payload_.data(), rank_payload_.size(),
        reinterpret_cast<const uint8_t*>(text.data()), text.size(),
        limit);
    if (r == -1) {
        throw EncodingError("turbotoken_is_within_token_limit_bpe_from_ranks failed");
    }
    if (r == -2) {
        return std::nullopt;
    }
    return static_cast<size_t>(r);
}

std::string Encoding::format_chat(
    const std::vector<ChatMessage>& messages,
    const ChatOptions& opts) const {
    auto tmpl = resolve_chat_template(opts.mode);
    std::string result;
    for (const auto& msg : messages) {
        result += tmpl.message_prefix;
        result += msg.role;
        result += '\n';
        result += msg.content;
        result += tmpl.message_suffix;
    }
    if (tmpl.assistant_prefix) {
        result += *tmpl.assistant_prefix;
    }
    return result;
}

std::vector<uint32_t> Encoding::encode_chat(
    const std::vector<ChatMessage>& messages,
    const ChatOptions& opts) const {
    return encode(format_chat(messages, opts));
}

size_t Encoding::count_chat(
    const std::vector<ChatMessage>& messages,
    const ChatOptions& opts) const {
    return count(format_chat(messages, opts));
}

std::optional<size_t> Encoding::is_chat_within_token_limit(
    const std::vector<ChatMessage>& messages,
    size_t limit,
    const ChatOptions& opts) const {
    return is_within_token_limit(format_chat(messages, opts), limit);
}

static std::string read_file_contents(const std::filesystem::path& path) {
    std::ifstream f(path, std::ios::binary);
    if (!f) {
        throw EncodingError("Cannot open file: " + path.string());
    }
    std::ostringstream ss;
    ss << f.rdbuf();
    return ss.str();
}

std::vector<uint32_t> Encoding::encode_file_path(
    const std::filesystem::path& path) const {
    auto path_str = path.string();
    ptrdiff_t n = turbotoken_encode_bpe_file_from_ranks(
        rank_payload_.data(), rank_payload_.size(),
        reinterpret_cast<const uint8_t*>(path_str.data()), path_str.size(),
        nullptr, 0);
    if (n < 0) {
        // Fallback: read file contents and encode directly
        auto contents = read_file_contents(path);
        return encode(contents);
    }

    std::vector<uint32_t> tokens(static_cast<size_t>(n));
    ptrdiff_t written = turbotoken_encode_bpe_file_from_ranks(
        rank_payload_.data(), rank_payload_.size(),
        reinterpret_cast<const uint8_t*>(path_str.data()), path_str.size(),
        tokens.data(), tokens.size());
    if (written < 0) {
        throw EncodingError("turbotoken_encode_bpe_file_from_ranks failed");
    }
    tokens.resize(static_cast<size_t>(written));
    return tokens;
}

size_t Encoding::count_file_path(const std::filesystem::path& path) const {
    auto path_str = path.string();
    ptrdiff_t n = turbotoken_count_bpe_file_from_ranks(
        rank_payload_.data(), rank_payload_.size(),
        reinterpret_cast<const uint8_t*>(path_str.data()), path_str.size());
    if (n < 0) {
        // Fallback: read file contents and count directly
        auto contents = read_file_contents(path);
        return count(contents);
    }
    return static_cast<size_t>(n);
}

std::optional<size_t> Encoding::is_file_path_within_token_limit(
    const std::filesystem::path& path, size_t limit) const {
    auto path_str = path.string();
    ptrdiff_t r = turbotoken_is_within_token_limit_bpe_file_from_ranks(
        rank_payload_.data(), rank_payload_.size(),
        reinterpret_cast<const uint8_t*>(path_str.data()), path_str.size(),
        limit);
    if (r == -1) {
        // Fallback: read file contents and check limit directly
        auto contents = read_file_contents(path);
        return is_within_token_limit(contents, limit);
    }
    if (r == -2) {
        return std::nullopt;
    }
    return static_cast<size_t>(r);
}

} // namespace turbotoken
